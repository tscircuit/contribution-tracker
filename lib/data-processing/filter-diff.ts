export function filterDiff(diff: string): string {
  if (!diff || typeof diff !== "string") {
    return "No diff available"
  }

  const lines = diff.split("\n")
  const filteredLines: string[] = []
  let skipFile = false
  let hasValidContent = false

  for (const line of lines) {
    if (line.startsWith("diff --git")) {
      const fileName = line.split(" ")[3]?.replace("b/", "") || ""
      skipFile =
        fileName.endsWith(".svg") ||
        fileName.endsWith(".lock") ||
        fileName.includes("package-lock.json") ||
        fileName.includes("yarn.lock") ||
        fileName.includes("pnpm-lock.yaml")

      if (!skipFile) {
        hasValidContent = true
      }
    }

    if (!skipFile) {
      filteredLines.push(line)
    }
  }

  const result = filteredLines.join("\n").trim()

  if (!result || !hasValidContent) {
    return "Changes only in non-code files (SVG, lock files, etc.)"
  }

  return result
}
