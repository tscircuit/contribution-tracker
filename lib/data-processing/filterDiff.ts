import parseDiff from "parse-diff"

export function filterDiff(diffContent: string): string {
  const filesToExclude = ["package-lock.json", "bun.lockb"]

  // Parse the diff content
  const files = parseDiff(diffContent)

  // Filter out the excluded files
  const filteredFiles = files.filter(
    (file) => file.to && !filesToExclude.includes(file.to),
  )

  // Reconstruct the diff content
  let result = ""

  for (const file of filteredFiles) {
    result += `diff --git a/${file.from} b/${file.to}\n`
    if (file.index) result += `index ${file.index}\n`
    if (file.deleted) {
      result += `--- a/${file.from}\n`
      result += "+++ /dev/null\n"
    } else if (file.new) {
      result += "--- /dev/null\n"
      result += `+++ b/${file.to}\n`
    } else {
      result += `--- a/${file.from}\n`
      result += `+++ b/${file.to}\n`
    }

    for (const chunk of file.chunks) {
      result += `@@ -${chunk.oldStart},${chunk.oldLines} +${chunk.newStart},${chunk.newLines} @@\n`
      for (const change of chunk.changes) {
        result += change.type + change.content + "\n"
      }
    }
  }

  return result.trim()
}

export default filterDiff
