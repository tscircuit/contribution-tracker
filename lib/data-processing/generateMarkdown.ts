import type { AnalyzedPR, ContributorStats } from "lib/types"
import { getContributorScore } from "../scoring"

export const impactIcon = (impact: "Major" | "Minor" | "Tiny") => {
  switch (impact) {
    case "Major":
      return "🐳 Major"
    case "Minor":
      return "🐙 Minor"
    case "Tiny":
      return "🐌 Tiny"
    default:
      return "🟣"
  }
}

export const scoreToStarString = (score: number) => {
  if (score <= 3) return ""
  if (score <= 10) return "⭐"
  if (score <= 30) return "⭐⭐"
  if (score <= 50) return "⭐⭐⭐"
  if (score <= 75) return "👑"
  if (score <= 100) return "👑👑"
  return "👑👑👑"
}

export async function generateMarkdown(
  prs: AnalyzedPR[],
  contributorIdToStatsMap: Record<string, ContributorStats>,
  weekStart: string,
): Promise<string> {
  let markdown = `# Contribution Overview ${weekStart}\n\n`

  // Generate Mermaid pie chart
  markdown += "## PRs by Repository\n\n"
  markdown += "```mermaid\npie\n"
  const repoCounts = prs.reduce(
    (acc, pr) => {
      acc[pr.repo] = (acc[pr.repo] || 0) + 1
      return acc
    },
    {} as Record<string, number>,
  )
  Object.entries(repoCounts).forEach(([repo, count]) => {
    markdown += `    "${repo}" : ${count}\n`
  })
  markdown += "```\n\n"

  // Generate contributor overview table
  markdown += "## Contributor Overview\n\n"

  markdown +=
    "| Contributor | 🐳 Major | 🐙 Minor | 🐌 Tiny | ⭐ | Issues Created | Discussion Contributions |\n"
  markdown +=
    "|-------------|---------|---------|---------|-----|----------------|--------------------------|\n"

  // Group PRs by contributor
  const prsByContributor = prs.reduce(
    (acc, pr) => {
      if (!acc[pr.contributor]) {
        acc[pr.contributor] = []
      }
      acc[pr.contributor].push(pr)
      return acc
    },
    {} as Record<string, AnalyzedPR[]>,
  )

  // Calculate scores for each contributor
  const contributorScores: Record<string, any> = {}

  Object.entries(prsByContributor).forEach(([contributor, contributorPRs]) => {
    // Get contributor stats if they exist
    const stats = contributorIdToStatsMap[contributor]

    // Calculate score
    const scoreResult = getContributorScore(contributorPRs, stats)

    // Store the result
    contributorScores[contributor] = {
      ...scoreResult,
      issuesCreated: stats?.issuesCreated ?? 0,
    }

    // Update the contributor's stats with the new score and counts
    if (stats) {
      stats.score = scoreResult.score
      stats.major = scoreResult.major
      stats.minor = scoreResult.minor
      stats.tiny = scoreResult.tiny
      stats.stars = scoreToStarString(scoreResult.score)
    }
  })

  // Sort contributors by score
  const sortedContributors = Object.entries(contributorScores).sort((a, b) => {
    return b[1].score - a[1].score
  })

  // Find contributors who exist in stats but not in sorted contributors
  const contributorsHavingOnlyDiscussions = Object.keys(
    contributorIdToStatsMap,
  ).filter(
    (contributor) =>
      !sortedContributors.some(
        ([sortedContributor]) => sortedContributor === contributor,
      ),
  )

  for (const contributor of contributorsHavingOnlyDiscussions) {
    // Add score for GitHub Discussions contributions
    // 1 point for each Participating comment (minor contribution)
    // 2 points for each Great Informative comment (major contribution)
    // 4 points for each Incredible comment (two major contributions)
    const discussionNormalComments =
      contributorIdToStatsMap[contributor].discussionNormalComments || 0
    const discussionGreatInformativeComments =
      contributorIdToStatsMap[contributor].discussionGreatInformativeComments ||
      0
    const discussionIncredibleComments =
      contributorIdToStatsMap[contributor].discussionIncredibleComments || 0

    if (!contributorIdToStatsMap[contributor].score) {
      contributorIdToStatsMap[contributor].score = 0
    }
  }

  // Generate table rows
  for (const [contributor, effort] of sortedContributors) {
    // Calculate discussion contributions summary
    const discussionNormalComments =
      contributorIdToStatsMap[contributor].discussionNormalComments || 0
    const discussionGreatInformativeComments =
      contributorIdToStatsMap[contributor].discussionGreatInformativeComments ||
      0
    const discussionIncredibleComments =
      contributorIdToStatsMap[contributor].discussionIncredibleComments || 0
    const discussionSummary = `${discussionNormalComments}🔹 ${discussionGreatInformativeComments}🔶 ${discussionIncredibleComments}💎`

    markdown += `| [${contributor}](#${contributor.replace(/\s/g, "-")}) | ${
      effort.major
    } | ${effort.minor} | ${effort.tiny} | ${scoreToStarString(
      effort.score,
    )} | ${effort.issuesCreated} | ${discussionSummary} |\n`
  }

  for (const contributor of contributorsHavingOnlyDiscussions) {
    const stats = contributorIdToStatsMap[contributor]
    if (!contributorIdToStatsMap[contributor].discussionComments) continue
    // Calculate discussion contributions summary
    const discussionNormalComments =
      contributorIdToStatsMap[contributor].discussionNormalComments || 0
    const discussionGreatInformativeComments =
      contributorIdToStatsMap[contributor].discussionGreatInformativeComments ||
      0
    const discussionIncredibleComments =
      contributorIdToStatsMap[contributor].discussionIncredibleComments || 0
    const discussionSummary = `${discussionNormalComments}🔹 ${discussionGreatInformativeComments}🔶 ${discussionIncredibleComments}💎`

    markdown += `| [${contributor}](#${contributor.replace(
      /\s/g,
      "-",
    )}) | 0 | 0 | 0 | ${scoreToStarString(stats.score || 0)} | ${
      stats?.issuesCreated ?? 0
    } | ${discussionSummary} |\n`
  }
  markdown += "\n"

  // Add explanation for discussion contribution symbols
  markdown += "### Discussion Contribution Legend\n\n"
  markdown += "- 🔹 Normal Comments: Basic participation with minimal effort\n"
  markdown +=
    "- 🔶 Great Informative Comments: Thoughtful participation that adds value\n"
  markdown +=
    "- 💎 Incredible Comments: Exceptional participation with high-quality content\n\n"

  // Generate Review Table
  markdown += "## Review Table\n\n"

  const columnTitleToDescription = {
    "Reviews Received":
      "Number of reviews received for PRs for this contributor",
    "Approvals Received":
      "Number of approvals received for PRs this contributor authored",
    "Rejections Received":
      "Number of rejections received for PRs this contributor authored",
    "PRs Opened": "Number of PRs opened by this contributor",
    "Issues Created": "Number of issues created by this contributor",
    "Bountied Issues":
      "Number of issues this contributor created with a bounty",
    "Bountied Issue $":
      "Total bounty amount placed on issues authored by this contributor",
  }

  markdown += Object.entries(columnTitleToDescription)
    .map(
      ([title, description]) =>
        `[${title
          .toLowerCase()
          .replace(/\s/g, "-")}-hover]: ## "${description}"`,
    )
    .join("\n")
  markdown += "\n\n"

  const columnTitleToPropName: Record<
    string,
    keyof ContributorStats | "contributor"
  > = {
    Contributor: "contributor",
    "Reviews Received": "reviewsReceived",
    "Approvals Received": "approvalsReceived",
    "Rejections Received": "rejectionsReceived",
    Approvals: "approvalsGiven",
    Rejections: "rejectionsGiven",
    "PRs Opened": "prsOpened",
    "PRs Merged": "prsMerged",
    "Issues Created": "issuesCreated",
    "Bountied Issues": "bountiedIssuesCount",
    "Bountied Issue $": "bountiedIssuesTotal",
  }

  const columnTitles = Object.keys(columnTitleToPropName)

  markdown += "|"
  columnTitles.forEach((column) => {
    markdown += ` ${column} |`
  })
  markdown += "\n"

  markdown += "|"
  columnTitles.forEach(() => {
    markdown += "---|"
  })
  markdown += "\n"

  Object.entries(contributorIdToStatsMap).forEach(
    ([contributor, stats]: any[]) => {
      markdown += "|"
      columnTitles.forEach((columnTitle) => {
        if (columnTitle.toLowerCase().trim() === "contributor") {
          markdown += ` [${contributor}](#${contributor.replace(/\s/g, "-")}) |`
          return
        }
        markdown += ` ${stats[columnTitleToPropName[columnTitle]]} |`
      })
      markdown += "\n"
    },
  )
  markdown += "\n"

  // Generate changes by repository
  markdown += "## Changes by Repository\n\n"
  const prsByRepo = prs.reduce(
    (acc, pr) => {
      if (!acc[pr.repo]) {
        acc[pr.repo] = []
      }
      acc[pr.repo].push(pr)
      return acc
    },
    {} as Record<string, AnalyzedPR[]>,
  )

  Object.entries(prsByRepo).forEach(([repo, repoPRs]) => {
    markdown += `### [${repo}](https://github.com/${repo})\n\n`
    markdown +=
      "| PR # | Impact | Contributor | Description | Milestone Aligned |\n"
    markdown +=
      "|------|--------|-------------|-------------|-------------------|\n"
    repoPRs
      .sort((a, b) => {
        const impactOrder = { Major: 0, Minor: 1, Tiny: 2 }
        return (
          impactOrder[a.impact as keyof typeof impactOrder] -
          impactOrder[b.impact as keyof typeof impactOrder]
        )
      })
      .forEach((pr) => {
        markdown += `| [#${pr.number}](${pr.url}) | ${impactIcon(
          pr.impact,
        )} | ${pr.contributor} | ${pr.description} | ${
          pr.isAlignedWithMilestone ? "✅" : "❌"
        } |\n`
      })
    markdown += "\n"
  })

  // Generate changes by contributor
  markdown += "## Changes by Contributor\n\n"

  Object.entries(prsByContributor).forEach(([contributor, contributorPRs]) => {
    markdown += `### [${contributor}](https://github.com/${contributor})\n\n`
    markdown += "| PR # | Impact | Description | Milestone Aligned |\n"
    markdown += "|------|--------|-------------|-------------------|\n"
    contributorPRs.forEach((pr) => {
      markdown += `| [#${pr.number}](${pr.url}) | ${impactIcon(pr.impact)} | ${
        pr.description
      } | ${pr.isAlignedWithMilestone ? "✅" : "❌"} |\n`
    })
    markdown += "\n"
  })

  // Generate Repository Owners Table
  markdown += "## Repository Owners\n\n"
  markdown += "| Repository | Codeowners |\n"
  markdown += "|------------|------------|\n"

  const repoOwnersWithPaths = Object.entries(contributorIdToStatsMap)
    .filter(([_, stats]) => stats.reposOwned && stats.reposOwned.length > 0)
    .reduce(
      (ownersMap, [ownerName, contributorStats]) => {
        contributorStats.reposOwned?.forEach(({ repo: repoName, paths }) => {
          if (!ownersMap[repoName]) {
            ownersMap[repoName] = { owners: [], paths: [] }
          }
          if (!ownersMap[repoName].owners.includes(ownerName)) {
            ownersMap[repoName].owners.push(ownerName)
          }
          if (paths && paths.length > 0) {
            paths.forEach((path) => {
              if (!ownersMap[repoName].paths.includes(path)) {
                ownersMap[repoName].paths.push(path)
              }
            })
          }
        })
        return ownersMap
      },
      {} as Record<string, { owners: string[]; paths: string[] }>,
    )

  Object.entries(repoOwnersWithPaths).forEach(([repo, { owners }]) => {
    const ownersLinks = owners
      .map((owner) => `[${owner}](https://github.com/${owner})`)
      .join(", ")
    const repoDisplayName = repo.replace(/^tscircuit\//, "")
    markdown += `| [${repoDisplayName}](https://github.com/${repo}/blob/main/.github/CODEOWNERS) | ${ownersLinks} |\n`
  })
  markdown += "\n"

  // Generate Repos by Owner Table
  markdown += "## Repos by Owner\n\n"
  markdown += "| User | Repo |\n"
  markdown += "|------|------|\n"

  const ownerRepos = Object.entries(contributorIdToStatsMap)
    .filter(([_, stats]) => stats.reposOwned && stats.reposOwned.length > 0)
    .reduce(
      (ownersMap, [ownerName, contributorStats]) => {
        if (!ownersMap[ownerName]) {
          ownersMap[ownerName] = []
        }
        contributorStats.reposOwned?.forEach(({ repo: repoName, paths }) => {
          if (!ownersMap[ownerName].some((repo) => repo.repo === repoName)) {
            ownersMap[ownerName].push({
              repo: repoName,
              paths: paths || [],
            })
          }
        })
        return ownersMap
      },
      {} as Record<string, Array<{ repo: string; paths: string[] }>>,
    )

  Object.entries(ownerRepos).forEach(([owner, repos]) => {
    repos.forEach((repoData, index) => {
      const userCell =
        index === 0 ? `[${owner}](https://github.com/${owner})` : ""
      const repoDisplayName = repoData.repo.replace(/^tscircuit\//, "")
      markdown += `| ${userCell} | [${repoDisplayName}](https://github.com/${repoData.repo}/blob/main/.github/CODEOWNERS) |\n`
    })
  })
  markdown += "\n"

  return markdown
}
