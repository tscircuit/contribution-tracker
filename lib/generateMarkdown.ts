import type { AnalyzedPR, ContributorStats } from "lib/types"

export const impactIcon = (impact: "Major" | "Minor" | "Tiny") => {
  switch (impact) {
    case "Major":
      return "🐳 Major"
    case "Minor":
      return "🐙 Minor"
    case "Tiny":
      return "🐌 Tiny"
    default:
      return "🟣"
  }
}

export const scoreToStarString = (score: number) => {
  if (score <= 3) return ""
  if (score <= 10) return "⭐"
  if (score <= 30) return "⭐⭐"
  if (score <= 50) return "⭐⭐⭐"
  if (score <= 75) return "👑"
  if (score <= 100) return "👑👑"
  return "👑👑👑"
}

export async function generateMarkdown(
  prs: AnalyzedPR[],
  contributorIdToStatsMap: Record<string, ContributorStats>,
  weekStart: string,
): Promise<string> {
  let markdown = `# Contribution Overview ${weekStart}\n\n`

  // Generate Mermaid pie chart
  markdown += "## PRs by Repository\n\n"
  markdown += "```mermaid\npie\n"
  const repoCounts = prs.reduce(
    (acc, pr) => {
      acc[pr.repo] = (acc[pr.repo] || 0) + 1
      return acc
    },
    {} as Record<string, number>,
  )
  Object.entries(repoCounts).forEach(([repo, count]) => {
    markdown += `    "${repo}" : ${count}\n`
  })
  markdown += "```\n\n"

  // Generate contributor overview table
  markdown += "## Contributor Overview\n\n"

  markdown +=
    "| Contributor | 🐳 Major | 🐙 Minor | 🐌 Tiny | ⭐ | Issues Created |\n"
  markdown +=
    "|-------------|---------|---------|---------|-----|----------------|\n"
  const impactWorth = { Major: 4, Minor: 2, Tiny: 1 }
  const contributorEffort = prs.reduce(
    (acc, pr) => {
      if (!acc[pr.contributor]) {
        acc[pr.contributor] = { Major: 0, Minor: 0, Tiny: 0, score: 0 }
      }
      acc[pr.contributor][pr.impact]++
      const impactScore = impactWorth[pr.impact as keyof typeof impactWorth]
      if (!Number.isNaN(impactScore) && typeof impactScore === "number") {
        acc[pr.contributor].score += impactScore
      }

      // Track number of issues created
      acc[pr.contributor].issuesCreated =
        contributorIdToStatsMap[pr.contributor]?.issuesCreated ?? 0 // Use fallback to 0

      return acc
    },
    {} as Record<string, Record<string, number>>,
  )

  Object.entries(contributorEffort).forEach(([contributor, effort]) => {
    const bountiedAmount =
      contributorIdToStatsMap[contributor]?.bountiedIssuesTotal || 0
    // Convert bounty amount to minor contributions ($10 = 1 minor contribution)
    let minorContributionsFromBounties = Math.floor(bountiedAmount / 10)
    // Cap at 10 minor contributions as per requirements
    minorContributionsFromBounties = Math.min(
      minorContributionsFromBounties,
      10,
    )
    // Add to score (minor contributions are worth 2 points each)
    effort.score += minorContributionsFromBounties * 2

    const approvalsGiven =
      contributorIdToStatsMap[contributor]?.approvalsGiven || 0
    const rejectionsGiven =
      contributorIdToStatsMap[contributor]?.rejectionsGiven || 0
    effort.score += Math.min(approvalsGiven + rejectionsGiven, 20)
  })

  const sortedContributors = Object.entries(contributorEffort).sort(
    (a, b) => b[1].score - a[1].score,
  )
  for (const [contributor, effort] of sortedContributors) {
    markdown += `| [${contributor}](#${contributor.replace(/\s/g, "-")}) | ${effort.Major} | ${effort.Minor} | ${effort.Tiny} | ${scoreToStarString(effort.score)} | ${effort.issuesCreated} |\n`

    // Update the contributor's stats with the new score
    contributorIdToStatsMap[contributor].score = effort.score
  }
  markdown += "\n"

  // Generate Review Table
  markdown += "## Review Table\n\n"

  const columnTitleToDescription = {
    "Reviews Received":
      "Number of reviews received for PRs for this contributor",
    "Approvals Received":
      "Number of approvals received for PRs this contributor authored",
    "Rejections Received":
      "Number of rejections received for PRs this contributor authored",
    "PRs Opened": "Number of PRs opened by this contributor",
    "Issues Created": "Number of issues created by this contributor",
    "Bountied Issues":
      "Number of issues this contributor created with a bounty",
    "Bountied Issue $":
      "Total bounty amount placed on issues authored by this contributor",
  }

  markdown += Object.entries(columnTitleToDescription)
    .map(
      ([title, description]) =>
        `[${title.toLowerCase().replace(/\s/g, "-")}-hover]: ## "${description}"`,
    )
    .join("\n")
  markdown += "\n\n"

  const columnTitleToPropName: Record<
    string,
    keyof ContributorStats | "contributor"
  > = {
    Contributor: "contributor",
    "Reviews Received": "reviewsReceived",
    "Approvals Received": "approvalsReceived",
    "Rejections Received": "rejectionsReceived",
    Approvals: "approvalsGiven",
    Rejections: "rejectionsGiven",
    "PRs Opened": "prsOpened",
    "PRs Merged": "prsMerged",
    "Issues Created": "issuesCreated",
    "Bountied Issues": "bountiedIssuesCount",
    "Bountied Issue $": "bountiedIssuesTotal",
  }

  const columnTitles = Object.keys(columnTitleToPropName)

  markdown += "|"
  columnTitles.forEach((column) => {
    markdown += ` ${column} |`
  })
  markdown += "\n"

  markdown += "|"
  columnTitles.forEach(() => {
    markdown += "---|"
  })
  markdown += "\n"

  Object.entries(contributorIdToStatsMap).forEach(
    ([contributor, stats]: any[]) => {
      markdown += "|"
      columnTitles.forEach((columnTitle) => {
        if (columnTitle.toLowerCase().trim() === "contributor") {
          markdown += ` [${contributor}](#${contributor.replace(/\s/g, "-")}) |`
          return
        }
        markdown += ` ${stats[columnTitleToPropName[columnTitle]]} |`
      })
      markdown += "\n"
    },
  )
  markdown += "\n"

  // Generate changes by repository
  markdown += "## Changes by Repository\n\n"
  const prsByRepo = prs.reduce(
    (acc, pr) => {
      if (!acc[pr.repo]) {
        acc[pr.repo] = []
      }
      acc[pr.repo].push(pr)
      return acc
    },
    {} as Record<string, AnalyzedPR[]>,
  )

  Object.entries(prsByRepo).forEach(([repo, repoPRs]) => {
    markdown += `### [${repo}](https://github.com/${repo})\n\n`
    markdown += "| PR # | Impact | Contributor | Description |\n"
    markdown += "|------|--------|-------------|-------------|\n"
    repoPRs
      .sort((a, b) => {
        const impactOrder = { Major: 0, Minor: 1, Tiny: 2 }
        return (
          impactOrder[a.impact as keyof typeof impactOrder] -
          impactOrder[b.impact as keyof typeof impactOrder]
        )
      })
      .forEach((pr) => {
        markdown += `| [#${pr.number}](${pr.url}) | ${impactIcon(
          pr.impact,
        )} | ${pr.contributor} | ${pr.description} |\n`
      })
    markdown += "\n"
  })

  // Generate changes by contributor
  markdown += "## Changes by Contributor\n\n"
  const prsByContributor = prs.reduce(
    (acc, pr) => {
      if (!acc[pr.contributor]) {
        acc[pr.contributor] = []
      }
      acc[pr.contributor].push(pr)
      return acc
    },
    {} as Record<string, AnalyzedPR[]>,
  )

  Object.entries(prsByContributor).forEach(([contributor, contributorPRs]) => {
    markdown += `### [${contributor}](https://github.com/${contributor})\n\n`
    markdown += "| PR # | Impact | Description |\n"
    markdown += "|------|--------|-------------|\n"
    contributorPRs.forEach((pr) => {
      markdown += `| [#${pr.number}](${pr.url}) | ${impactIcon(pr.impact)} | ${
        pr.description
      } |\n`
    })
    markdown += "\n"
  })

  return markdown
}
