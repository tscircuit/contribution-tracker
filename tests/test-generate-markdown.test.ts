import { describe, it, expect } from "bun:test"
import { generateMarkdown } from "lib/data-processing/generateMarkdown"
import type { AnalyzedPR, ContributorStats } from "lib/types"

const minimalPrAttributes = {
  mostly_style: false,
  new_page_or_component: false,
  introduces_or_fixes_a_footprint: false,
  core_change: false,
  only_dependency_update: false,
  bad_title: false,
  introduces_new_circuit_board: false,
  fixes_circuit_board: false,
  fixes_subtle_important_bug: false,
  minor_fix: false,
  major_autorouter_bug_fix: false,
  only_adds_autorouter_fixtures: false,
  only_reproduces_a_bug: false,
  reproduces_and_fixes_a_bug: false,
  minor_developer_experience_improvement: false,
  major_experience_improvement: false,
  introduces_new_schematic_symbol: false,
  fixes_schematic_representation: false,
  improves_parts_engine: false,
  add_design_to_schematic_corpus: false,
  major_improvement_to_core_data_modeling: false,
  major_library_algorithm_contribution: false,
  substantially_improves_svg_generation: false,
}

const mockPRs: AnalyzedPR[] = [
  {
    ...minimalPrAttributes,
    number: 1,
    state: "merged",
    title: "Add feature X",
    body: "Implements feature X.",
    user: { login: "alice" },
    html_url: "https://github.com/org/repo/pull/1",
    created_at: "2024-06-01T00:00:00Z",
    merged_at: "2024-06-02T00:00:00Z",
    description: "Add feature X",
    impact: "Major",
    contributor: "alice",
    repo: "org/repo",
    url: "https://github.com/org/repo/pull/1",
    isAlignedWithMilestone: true,
    starRating: 3,
    mostly_style: true,
    core_change: true,
  },
  {
    ...minimalPrAttributes,
    number: 2,
    state: "merged",
    title: "Fix bug Y",
    body: "Fixes bug Y.",
    user: { login: "bob" },
    html_url: "https://github.com/org/repo/pull/2",
    created_at: "2024-06-03T00:00:00Z",
    merged_at: "2024-06-04T00:00:00Z",
    description: "Fix bug Y",
    impact: "Minor",
    contributor: "bob",
    repo: "org/repo",
    url: "https://github.com/org/repo/pull/2",
    isAlignedWithMilestone: false,
    starRating: 2,
    minor_fix: true,
  },
  {
    ...minimalPrAttributes,
    number: 3,
    state: "merged",
    title: "Update docs",
    body: "Updates documentation.",
    user: { login: "alice" },
    html_url: "https://github.com/org/another-repo/pull/3",
    created_at: "2024-06-05T00:00:00Z",
    merged_at: "2024-06-06T00:00:00Z",
    description: "Update docs",
    impact: "Tiny",
    contributor: "alice",
    repo: "org/another-repo",
    url: "https://github.com/org/another-repo/pull/3",
    isAlignedWithMilestone: false,
    starRating: 1,
    bad_title: true,
  },
]

const mockStats: Record<string, ContributorStats> = {
  alice: {
    reviewsReceived: 1,
    rejectionsReceived: 0,
    approvalsReceived: 1,
    approvalsGiven: 2,
    rejectionsGiven: 0,
    prsOpened: 2,
    prsMerged: 1,
    issuesCreated: 1,
    bountiedIssuesCount: 0,
    bountiedIssuesTotal: 0,
    discussionNormalComments: 2,
    discussionGreatInformativeComments: 1,
    discussionIncredibleComments: 0,
    reposOwned: [{ repo: "org/repo", paths: ["src/"] }],
  },
  bob: {
    reviewsReceived: 0,
    rejectionsReceived: 0,
    approvalsReceived: 0,
    approvalsGiven: 1,
    rejectionsGiven: 0,
    prsOpened: 1,
    prsMerged: 0,
    issuesCreated: 0,
    bountiedIssuesCount: 0,
    bountiedIssuesTotal: 0,
    discussionNormalComments: 0,
    discussionGreatInformativeComments: 0,
    discussionIncredibleComments: 1,
    reposOwned: [],
  },
}

describe("generateMarkdown", () => {
  it("should generate markdown with expected sections and contributor data", async () => {
    const markdown = await generateMarkdown(mockPRs, mockStats, "2024-07-01", {
      "tscircuit/gyromug": ["alice", "bob"],
      "tscircuit/schematic-corpus": ["Abse2001", "alice"],
      "tscircuit/lol": []
    })
    expect(markdown).toContain("# Contribution Overview 2024-07-01")
    expect(markdown).toContain("## Contributor Overview")
    expect(markdown).toContain("alice")
    expect(markdown).toContain("bob")
    expect(markdown).toContain("org/repo")
    expect(markdown).toContain("org/another-repo")
    expect(markdown).toContain("Add feature X")
    expect(markdown).toContain("Fix bug Y")
    expect(markdown).toContain("Update docs")
    expect(markdown).toContain("Discussion Contribution Legend")
    expect(markdown).toContain("Repository Owners")
    expect(markdown).toContain("Repositories by Owner")
    expect(markdown).toMatchSnapshot("markdown")
  })
})
