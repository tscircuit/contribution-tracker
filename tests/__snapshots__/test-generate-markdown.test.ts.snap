// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`generateMarkdown should generate markdown with expected sections and contributor data: markdown 1`] = `
"# Contribution Overview 2024-07-01

## PRs by Repository

\`\`\`mermaid
pie
    "org/repo" : 2
    "org/another-repo" : 1
\`\`\`

## Contributor Overview

| Contributor | 🐳 Major | 🐙 Minor | 🐌 Tiny | ⭐ | Issues Created | Discussion Contributions |
|-------------|---------|---------|---------|-----|----------------|--------------------------|
| [alice](#alice) | 1 | 0 | 1 | ⭐ | 1 | 2🔹 1🔶 0💎 |
| [bob](#bob) | 0 | 1 | 0 |  | 0 | 0🔹 0🔶 1💎 |

### Discussion Contribution Legend

- 🔹 Normal Comments: Basic participation with minimal effort
- 🔶 Great Informative Comments: Thoughtful participation that adds value
- 💎 Incredible Comments: Exceptional participation with high-quality content

## Review Table

[reviews-received-hover]: ## "Number of reviews received for PRs for this contributor"
[approvals-received-hover]: ## "Number of approvals received for PRs this contributor authored"
[rejections-received-hover]: ## "Number of rejections received for PRs this contributor authored"
[prs-opened-hover]: ## "Number of PRs opened by this contributor"
[issues-created-hover]: ## "Number of issues created by this contributor"
[bountied-issues-hover]: ## "Number of issues this contributor created with a bounty"
[bountied-issue-$-hover]: ## "Total bounty amount placed on issues authored by this contributor"

| Contributor | Reviews Received | Approvals Received | Rejections Received | Approvals | Rejections | PRs Opened | PRs Merged | Score | Issues Created | Bountied Issues | Bountied Issue $ |
|---|---|---|---|---|---|---|---|---|---|---|---|
| [alice](#alice) | 1 | 1 | 0 | 2 | 0 | 2 | 1 | 5 | 1 | 0 | 0 |
| [bob](#bob) | 0 | 0 | 0 | 1 | 0 | 1 | 0 | 2 | 0 | 0 | 0 |

## Top 7 Repositories by Contribution Points

\`\`\`mermaid
pie
    "org/repo" : 6
    "org/another-repo" : 1
\`\`\`

## Changes by Repository

### [org/repo](https://github.com/org/repo)

| PR # | Impact | Rating | Contributor | Description |
|------|--------|--------|-------------|-------------|
| [#1](https://github.com/org/repo/pull/1) | 🐳 Major | ⭐⭐⭐ | alice | Add feature X |
| [#2](https://github.com/org/repo/pull/2) | 🐙 Minor | ⭐⭐ | bob | Fix bug Y |

### [org/another-repo](https://github.com/org/another-repo)


<details>
<summary>🐌 Tiny Contributions (1)</summary>

| PR # | Impact | Contributor | Description |
|------|--------|-------------|-------------|
| [#3](https://github.com/org/another-repo/pull/3) | 🐌 Tiny | alice | Update docs |

</details>

## Changes by Contributor

### [alice](https://github.com/alice)

| PRs # | Impact | Rating | Description |
|------|--------|--------|-------------|
| [#1](https://github.com/org/repo/pull/1) | 🐳 Major | ⭐⭐⭐ | Add feature X |

<details>
<summary>🐌 Tiny Contributions (1)</summary>

| PR # | Impact | Description |
|------|--------|-------------|
| [#3](https://github.com/org/another-repo/pull/3) | 🐌 Tiny | Update docs |

</details>

### [bob](https://github.com/bob)

| PRs # | Impact | Rating | Description |
|------|--------|--------|-------------|
| [#2](https://github.com/org/repo/pull/2) | 🐙 Minor | ⭐⭐ | Fix bug Y |

## Repository Owners

| Repository | Codeowners |
|------------|------------|

## Repositories by Owner

| User | Repo |
|------|------|

"
`;
